kame/kame/pmsft/pmsft.1
.\" Copyright (C) 1995, 1996, 1997, 1998, and 1999 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Aug 10, 2004
.Dt PMSFT 1
.Os KAME
.\"
.Sh NAME
.Nm pmsft
.Nd protocol-independent Multicast Source Filter Testing Tool
.\"
.Sh SYNOPSIS
.Nm
.Op Ar file
.\"
.Sh DESCRIPTION
The 
.Nm
utility specifies a Multicast Source Filter (MSF) using APIs 
specified in Socket Interface Extention for MSF (RFC3678).
Since
.Nm
adopts a protocol-independent API, both IPv4 and IPv6 MSF are
supported.
.Pp
.Nm 
specifies an MSF on command-prompt "pmsft>".
Here is the list of commands that can be specified:

.Bl -tag -width -compact
.It Cm j Ar grp Ar ifname
join a group
.Ar grp
on interface
.Ar ifname.

.It Cm l Ar group Ar ifname
leave a group
.Ar grp
on interface
.Ar ifname.

.It Cm i Ar grp Ar ifname Ar num
set an include mode MSF for group
.Ar grp
on interface
.Ar ifname.
(followed by 
.Ar num
lines of source addresses)

.It Cm e Ar grp Ar ifname Ar num
set an exclude mode MSF for group
.Ar grp
on interface
.Ar ifname.
(followed by 
.Ar num
lines of source addresses)

.It Cm t Ar grp Ar ifname Ar src
accept multicast packets from
.Ar src
to
.Ar grp
on interface
.Ar ifname

.It Cm b Ar grp Ar ifname Ar src
deny multicast packets from
.Ar src
to
.Ar grp
on interface
.Ar ifname


.It Cm g Ar grp Ar ifname Ar num
get and show MSF regarding group
.Ar grp
on interface
.Ar ifname
(atmost 
.Ar num
sources).

.It Cm s Ar seconds
sleep for the specified
.Ar seconds.

.It Cm f Ar filename
read command(s) from file
.Ar filename.

.It Cm q
quit
.Nm
.El

.Pp
If there is an optional argument
.Ar file
, it is equivalent to the following command:
.Bd -literal
pmsft> f \fIfile\fR
.Ed
.
.Sh HISTORY
The
.Nm
command is originally based on wilbertdg@hetnet.nl's pmsft, supporting 
only IGMPv3.


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/pmsft/pmsft.c
/*
 * $KAME: pmsft.c,v 1.1 2004/08/09 02:04:36 suz Exp $
 * pmsft - test a IGMPv3/MLDv2 host stack, using protocol independent API
 * based on wilbertdg@hetnet.nl's pmsft, supporting only IGMPv3
 */

#define MAX_ADDRS		500

#include <ctype.h>
#include <err.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/errno.h>
#include <sys/socket.h>
#include <netdb.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <netinet/in.h>

int s = -1;

/*
 * Prototypes
 */
void process_file(char *);
void process_cmd(char*, FILE *fp);
void usage();
int comp_sas(const void *, const void *);

int
main(int argc, char **argv)
{
	int i;
	char line[BUFSIZ], *p;

	/* Process commands until user wants to quit */
	while (1) {
		printf("pmsft> ");
		if (fgets(line, sizeof(line), stdin) == NULL)
			return 0;

		if (line[0] != 'f') {
			process_cmd(line, stdin);
			continue;
		}

		/* process the given file */
		for (i = 1; isblank(line[i]); i++)
			;
		if ((p = (char*) strchr(line, '\n')) != NULL)
			*p = '\0';
		process_file(&line[i]);
	}
}

/*
 * Process commands from a file
 */
void
process_file(char *fname)
{
	char *lineptr;
	char line[BUFSIZ];
	FILE *fp;

	/* Try to open the file and feed all commands to process_cmd() */
	if ((fp = fopen(fname, "r")) == NULL) {
		errx(1, "fopen");
	}

	while (fgets(line, sizeof(line), fp) != NULL) {
		/* Skip comments and empty lines */
		lineptr = line;
		while (isblank(*lineptr)) 
			lineptr++;
		if (*lineptr != '#' && *lineptr != '\n')
			process_cmd(lineptr, fp);
	}
	fclose(fp);
	return;
}

/*
 * Process a line/command
 */
void
process_cmd(char *cmd, FILE *fp)
{
	struct group_req gr;
	struct group_source_req gsr;
	struct group_filter *gfp;
	struct addrinfo hints, *res, *res0;
	int error;
	char buffer[GROUP_FILTER_SIZE(MAX_ADDRS)], *p;
	char str1[NI_MAXHOST], str2[NI_MAXHOST], *line, ifname[IFNAMSIZ];
	int i, n, opt, level;

	/* Skip whitespaces */
	line = cmd + 1;
	while (isblank(*line)) 
		line++;

	switch (*cmd) {
	case '?':
		/* Show usage */
		usage();
		break;

	case 'q':
		/* Quit */
		if (s > 0)
			close(s);
		exit(0);

	case 's':
		/* Wait for some time */
		if (sscanf(line, "%d", &n) != 1) {
			warnx("invalid format");
			break;
		}
		if (n < 1) {
			warnx("invalid value");
			break;
		}
		for (i = 0; i < n; i++)
			sleep(1);
		break;

	case 'j':
	case 'l':
		/* Join or leave a multicast group */
		if (sscanf(line, "%s %s", str1, ifname) != 2) {
			warnx("invalid format"); 
			break;
		}

		memset(&hints, 0, sizeof(hints));
		hints.ai_family = PF_UNSPEC;
		hints.ai_socktype = SOCK_DGRAM;
		hints.ai_flags = AI_NUMERICHOST;
		error = getaddrinfo(str1, NULL, &hints, &res0);
		if (error) {
			warnx("%s", gai_strerror(error));
			break;
		}
		for (res = res0; res; res = res->ai_next) {
			if (s < 0)
				s = socket(res->ai_family, res->ai_socktype,
					   res->ai_protocol);
			if (s < 0)
				continue;
			break;  /* okay we got one */
		}
		memcpy(&gr.gr_group, res->ai_addr, res->ai_addrlen);
		if ((gr.gr_interface = if_nametoindex(ifname)) == 0) {
			perror("if_nametoindex");
			break;
		}
		opt = (*cmd == 'j') ? MCAST_JOIN_GROUP : MCAST_LEAVE_GROUP;
		level = gr.gr_group.ss_family == AF_INET6 ? IPPROTO_IPV6 :
			gr.gr_group.ss_family == AF_INET ? IPPROTO_IP : 0;
		if (setsockopt(s, level, opt, &gr, sizeof(gr)) == -1)
			perror("MCAST_JOIN/LEAVE_GROUP");
		freeaddrinfo(res0);
		break;

	case 'i':
	case 'e':
		/* 
		 * Set the socket to include or exclude filter mode, and
		 * add some sources to the filterlist, using the full-state,
		 * or advanced api 
		 */
		if (sscanf(line, "%s %s %d", str1, ifname, &n) != 3) {
			warnx("invalid format");
			break;
		}
		if (n > MAX_ADDRS) {
			warnx("invalid value");
			break;
		}

		memset(&hints, 0, sizeof(hints));
		hints.ai_family = PF_UNSPEC;
		hints.ai_socktype = SOCK_DGRAM;
		hints.ai_flags = AI_NUMERICHOST;
		error = getaddrinfo(str1, NULL, &hints, &res0);
		if (error) {
			warnx("%s", gai_strerror(error));
			break;
		}
		for (res = res0; res; res = res->ai_next) {
			if (s < 0)
				s = socket(res->ai_family, res->ai_socktype,
					   res->ai_protocol);
			if (s < 0)
				continue;
			break;  /* okay we got one */
		}

		/* Prepare argument */
		gfp = (struct group_filter*) buffer;
		memcpy(&gfp->gf_group, res->ai_addr, res->ai_addrlen);
		gfp->gf_fmode = (*cmd == 'i') ? MCAST_INCLUDE : MCAST_EXCLUDE;
		if ((gfp->gf_interface = if_nametoindex(ifname)) == 0) {
			perror("if_nametoindex");
			break;
		}
		gfp->gf_numsrc = n;
		freeaddrinfo(res0);

		for (i = 0; i < n; i++) {
			fgets(str1, sizeof(str1), fp);
			if ((p = (char*) strchr(str1, '\n')) != NULL)
				*p = '\0';

			memset(&hints, 0, sizeof(hints));
			hints.ai_family = PF_UNSPEC;
			hints.ai_socktype = SOCK_DGRAM;
			hints.ai_flags = AI_NUMERICHOST;
			error = getaddrinfo(str1, NULL, &hints, &res0);
			if (error) {
				warnx("%s", gai_strerror(error));
				break;
			}
			for (res = res0; res; res = res->ai_next) {
				break;  /* okay we got one */
			}
			memcpy(&gfp->gf_slist[i], res->ai_addr, res->ai_addrlen);
			freeaddrinfo(res0);
		}
		/* Execute ioctl() */
		if (ioctl(s, SIOCSMSFILTER, (void*) gfp) != 0) {
			perror("SIOCSMSFILTER");
		}
		break;

	case 't':
	case 'b':
		/* Allow or block traffic from a source, using the delta based api */
		if (sscanf(line, "%s %s %s", str1, ifname, str2) != 3) {
			warnx("invalid format");
			break;
		}
		gfp = (struct group_filter*) buffer;
		if ((gfp->gf_interface = if_nametoindex(ifname)) == 0) {
			perror("if_nametoindex");
			break;
		}
		gfp->gf_numsrc = 0;

		memset(&hints, 0, sizeof(hints));
		hints.ai_family = PF_UNSPEC;
		hints.ai_socktype = SOCK_DGRAM;
		hints.ai_flags = AI_NUMERICHOST;
		error = getaddrinfo(str1, NULL, &hints, &res0);
		if (error) {
			warnx("%s", gai_strerror(error));
			break;
		}
		for (res = res0; res; res = res->ai_next) {
			if (s < 0)
				s = socket(res->ai_family, res->ai_socktype,
					   res->ai_protocol);
			if (s < 0)
				continue;
			break;  /* okay we got one */
		}
		memcpy(&gfp->gf_group, res->ai_addr, res->ai_addrlen);
		freeaddrinfo(res0);

		/* First find out current filter mode */
		if (ioctl(s, SIOCGMSFILTER, gfp) != 0) {
			/* 
			 * It's only okay for 't' to fail, since the operation
			 * MCAST_JOIN_SOURCE_GROUP on a non existing membership
			 * should result in a new membership
			 */
			if (*cmd != 't') {
				perror("SIOCGMSFILTER");
				break;
			}
			gfp->gf_fmode = MCAST_INCLUDE;
		}
		if (gfp->gf_fmode == MCAST_EXCLUDE) {
			/* Any source */
			opt = (*cmd == 't') ? 
				MCAST_UNBLOCK_SOURCE : MCAST_BLOCK_SOURCE;
		} else {
			/* Controlled source */
			opt = (*cmd == 't') ? 
				MCAST_JOIN_SOURCE_GROUP : MCAST_LEAVE_SOURCE_GROUP;
		}

		/* Prepare argument */
		memset(&hints, 0, sizeof(hints));
		hints.ai_family = PF_UNSPEC;
		hints.ai_socktype = SOCK_DGRAM;
		hints.ai_flags = AI_NUMERICHOST;
		error = getaddrinfo(str2, NULL, &hints, &res0);
		if (error) {
			warnx("%s", gai_strerror(error));
			break;
		}
		for (res = res0; res; res = res->ai_next) {
			break;  /* okay we got one */
		}
		memcpy(&gsr.gsr_source, res->ai_addr, res->ai_addrlen);
		freeaddrinfo(res0);

		if ((gsr.gsr_interface = if_nametoindex(ifname)) == 0) {
			perror("if_nametoindex");
			break;
		}
		memcpy(&gsr.gsr_group, &gfp->gf_group, gfp->gf_group.ss_len);

		/* Execute setsockopt() */ 
		level = gsr.gsr_group.ss_family == AF_INET6 ? IPPROTO_IPV6 :
			gsr.gsr_group.ss_family == AF_INET ? IPPROTO_IP : 0;
		if (setsockopt(s, level, opt, &gsr, sizeof(gsr)) == -1)
			perror("setsockopt");	
		break;

	case 'g':
		/* Get and show the current filter mode, and the sources in the list */
		if (sscanf(line, "%s %s %d", str1, ifname, &n) != 3) {
			warnx("invalid format");
			break;
		}

		/* Prepare argument */
		gfp = (struct group_filter*) buffer;
		if ((gfp->gf_interface = if_nametoindex(ifname)) == 0) {
			perror("if_nametoindex");
			break;
		}
		gfp->gf_numsrc = n;

		memset(&hints, 0, sizeof(hints));
		hints.ai_family = PF_UNSPEC;
		hints.ai_socktype = SOCK_DGRAM;
		hints.ai_flags = AI_NUMERICHOST;
		error = getaddrinfo(str1, NULL, &hints, &res0);
		if (error) {
			warnx("%s", gai_strerror(error));
			break;
		}
		for (res = res0; res; res = res->ai_next) {
			if (s < 0)
				s = socket(res->ai_family, res->ai_socktype,
					   res->ai_protocol);
			if (s < 0)
				continue;
			break;  /* okay we got one */
		}
		memcpy(&gfp->gf_group, res->ai_addr, res->ai_addrlen);
		freeaddrinfo(res0);

		/* Execute ioctl() */
		if (ioctl(s, SIOCGMSFILTER, gfp) != 0) {
			perror("SIOCGMSFILTER");
			break;
		}

		printf("%s\n",
		       gfp->gf_fmode == MCAST_INCLUDE ? "include" : "exclude");
		if (n > gfp->gf_numsrc) {
			n = gfp->gf_numsrc;
			qsort(gfp->gf_slist, n,
			      sizeof(struct sockaddr_storage), &comp_sas);
			for (i = 0; i < n; i++) {
				char hbuf[NI_MAXHOST];
				struct sockaddr *sa;
				
				sa = (struct sockaddr *)&gfp->gf_slist[i];

				error = getnameinfo(sa, sa->sa_len,
						    hbuf, sizeof(hbuf),
						    NULL, 0, NI_NUMERICHOST);
				if (error) {
					warnx("%s", gai_strerror(error));
					continue;
				}
				printf("%s\n", hbuf);
			}
		}
		break;

	case '\n':
	default:
		break;
	}
}

/*
 * Print usage information
 */
void usage()
{
	printf("j group-addr ifname          "
		"- join IP multicast group\n");
	printf("l group-addr ifname          "
		"- leave IP multicast group\n");
	printf("i group-addr ifname n        "
		"- set n include mode src filters "
		"(followed by n lines of src-addrs)\n");
	printf("e group-addr ifname n        "
		"- set n exclude mode src filters "
		"(followed by n lines of src-addrs)\n");
	printf("t group-addr ifname src-addr "
		"- allow traffic from src\n");
	printf("b group-addr ifname src-addr "
		"- block traffic from src\n");
	printf("g group-addr ifname n        "
		"- get and show (max n) src filters\n");
	printf("f filename                   "
		"- read command(s) from file\n");
	printf("s seconds                    "
		"- sleep for some time\n");
	printf("q                            "
		"- quit\n");
}

int 
comp_sas(const void *a, const void *b)
{
	struct sockaddr *sa, *sb;
	char ha[NI_MAXHOST], hb[NI_MAXHOST];
	sa = (struct sockaddr *) a;
	sb = (struct sockaddr *) b;

	getnameinfo(sa, sa->sa_len, ha, sizeof(ha), NULL, 0, NI_NUMERICHOST);
	getnameinfo(sb, sb->sa_len, hb, sizeof(hb), NULL, 0, NI_NUMERICHOST);
	return strncmp(ha, hb, sizeof(ha));
}
